man/man1/turnserver.1
.\" Text automatically generated by txt2man
.TH TURN 1 "15 June 2015" "" ""
.SH GENERAL INFORMATION

The \fBTURN Server\fP project contains the source code of a TURN server and TURN client 
messaging library. Also, some extra programs provided, for testing\-only 
purposes. 
.PP
See the INSTALL file for the building instructions.
.PP
After the build, you will have the following binary images:
.TP
.B
1.
\fIturnserver\fP: \fBTURN Server\fP relay. 
The compiled binary image of the \fBTURN Server\fP program is located in bin/ sub\-directory.
.TP
.B
2.
\fIturnadmin\fP: TURN administration tool. See README.turnadmin and \fIturnadmin\fP man page.
.TP
.B
3.
turnutils_uclient. See README.turnutils and \fIturnutils\fP man page.
.TP
.B
4.
turnutils_peer. See README.turnutils and \fIturnutils\fP man page.
.TP
.B
5.
turnutils_stunclient. See README.turnutils and \fIturnutils\fP man page.
.TP
.B
6.
turnutils_rfc5769check. See README.turnutils and \fIturnutils\fP man page.
.PP
In the "examples/scripts" sub\-directory, you will find the examples of command lines to be used 
to run the programs. The scripts are meant to be run from examples/ sub\-directory, for example:
.PP
$ cd examples
$ ./scripts/secure_relay.sh
.SH RUNNING THE TURN SERVER

Options note: \fIturnserver\fP has long and short option names, for most options.
Some options have only long form, some options have only short form. Their syntax 
somewhat different, if an argument is required:
.PP
The short form must be used as this (for example):
.PP
.nf
.fam C
  $ turnserver \-L 12.34.56.78

.fam T
.fi
The long form equivalent must use the "=" character:
.PP
.nf
.fam C
  $ turnserver \-\-listening\-ip=12.34.56.78

.fam T
.fi
If this is a flag option (no argument required) then their usage are the same, for example:
.PP
.nf
.fam C
 $ turnserver \-a

.fam T
.fi
is equivalent to:
.PP
.nf
.fam C
 $ turnserver \-\-lt\-cred\-mech

.fam T
.fi
=====================================
.SS  NAME
\fB
\fBturnserver \fP\- a TURN relay server implementation.
\fB
.SS  SYNOPSIS
.nf
.fam C

$ \fIturnserver\fP [\fB\-n\fP | \fB\-c\fP <config\-file> ] [\fIflags\fP] [ \fB\-\-userdb\fP=<userdb\-file> | \fB\-\-psql\-userdb\fP=<db\-conn\-string> | \fB\-\-mysql\-userdb\fP=<db\-conn\-string>  | \fB\-\-redis\-userdb\fP=<db\-conn\-string> ] [\fB\-z\fP | \fB\-\-no\-auth\fP | \fB\-a\fP | \fB\-\-lt\-cred\-mech\fP ] [\fIoptions\fP]
$ \fIturnserver\fP \fB\-h\fP

.fam T
.fi
.fam T
.fi
.SS  DESCRIPTION                                           

.TP
.B
Config file settings:
.TP
.B
\fB\-n\fP
Do not use configuration file, use only command line parameters.
.TP
.B
\fB\-c\fP
Configuration file name (default \- turnserver.conf).
The format of config file can be seen in
the supplied examples/etc/turnserver.conf example file. Long 
names of the \fIoptions\fP are used as the configuration 
items names in the file. If not an absolute path is supplied, 
then the file is searched in the following directories: 
.RS
.IP \(bu 3
current directory
.IP \(bu 3
current directory etc/ sub\-directory
.IP \(bu 3
upper directory level etc/
.IP \(bu 3
/etc/
.IP \(bu 3
/usr/local/etc/
.IP \(bu 3
installation directory /etc
.RE
.TP
.B
User database settings:
.TP
.B
\fB\-b\fP, \fB\-\-userdb\fP
User database file name (default \- turnuserdb.conf),
for long\-term credentials mechanism only.
This user file database is being dynamically checked while the \fIturnserver\fP 
is working, and the user accounts can be changed dynamically by
editing the database.
.TP
.B
\fB\-e\fP, \fB\-\-psql\-userdb\fP
User database connection string for PostgreSQL.
This database can be used for long\-term and short\-term credentials mechanisms,
and it can store the secret value for secret\-based timed authentication in TURN RESP API.
The connection string format is like that:
.RS
.PP
"host=<host> dbname=<dbname> user=<db\-user> password=<db\-user\-password> connect_timeout=<seconds>" 
(for 8.x or newer Postgres).
.PP
Or:
.PP
"postgresql://username:password@hostname:port/databasename" (for 9.x or newer Postgres). 
See the INSTALL file for more explanations and examples.
.PP
Also, see http://www.PostgreSQL.org for full PostgreSQL documentation.
.RE
.TP
.B
\fB\-M\fP, \fB\-\-mysql\-userdb\fP
User database connection string for MySQL or MariaDB. 
This database can be used for long\-term and short\-term credentials mechanisms,
and it can store the secret value for secret\-based timed authentication in TURN RESP API.
The connection string format is like that:
.RS
.PP
"host=<host> dbname=<dbname> user=<db\-user> password=<db\-user\-password> connect_timeout=<seconds>"
See the INSTALL file for more explanations and examples.
.PP
Also, see http://www.mysql.org or http://mariadb.org 
for full MySQL documentation.
.PP
Optional connection string parameters for the secure communications (SSL): 
ca, capath, cert, key, cipher 
(see http://dev.mysql.com/doc/refman/5.1/en/ssl\-options.html for the 
command \fIoptions\fP description).
.RE
.TP
.B
\fB\-N\fP, \fB\-\-redis\-userdb\fP
User database connection string for Redis. 
This database can be used for long\-term and short\-term credentials mechanisms,
and it can store the secret value for secret\-based timed authentication in TURN RESP API.
The connection string format is like that:
.RS
.PP
"ip=<ip\-addr> dbname=<db\-number> password=<db\-password> connect_timeout=<seconds>"
See the INSTALL file for more explanations and examples.
.PP
Also, see http://redis.io for full Redis documentation.
.RE
.TP
.B
Flags:
.TP
.B
\fB\-v\fP, \fB\-\-verbose\fP
Moderate verbose mode.
.TP
.B
\fB\-V\fP, \fB\-\-Verbose\fP
Extra verbose mode, very annoying and not recommended.
.TP
.B
\fB\-o\fP, \fB\-\-daemon\fP
Run server as daemon.
.TP
.B
\fB\-f\fP, \fB\-\-fingerprint\fP
Use fingerprints in the TURN messages. If an incoming request
contains a fingerprint, then TURN server will always add 
fingerprints to the messages in this session, regardless of the
per\-server setting.
.TP
.B
\fB\-a\fP, \fB\-\-lt\-cred\-mech\fP
Use long\-term credentials mechanism (this one you need for WebRTC usage). 
This option can be used with either flat file user database or 
PostgreSQL DB or MySQL DB or Redis for user keys storage.
.TP
.B
\fB\-A\fP, \fB\-\-st\-cred\-mech\fP
Use the short\-term credentials mechanism. This option requires
a PostgreSQL or MySQL or Redis DB for short term passwords storage.
.TP
.B
\fB\-z\fP, \fB\-\-no\-auth\fP
Do not use any credentials mechanism, allow anonymous access. 
Opposite to \fB\-a\fP and \fB\-A\fP \fIoptions\fP. This is default option when no 
authentication\-related \fIoptions\fP are set.
By default, no credential mechanism is used \-
any user is allowed.
.TP
.B
\fB\-\-use\-auth\-secret\fP
TURN REST API flag.
Flag that sets a special WebRTC authorization option 
that is based upon authentication secret. The feature purpose 
is to support "\fBTURN Server\fP REST API" as described in
the TURN REST API section below.
This option uses timestamp as part of combined username:
usercombo \-> "timestamp:username",
turn user \-> usercombo,
turn password \-> \fBbase64\fP(hmac(input_buffer = usercombo, key = shared\-secret)).
This allows TURN credentials to be accounted for a specific user id.
If you don't have a suitable id, the timestamp alone can be used.
This option is just turns on secret\-based authentication.
The actual value of the secret is defined either by option static\-auth\-secret,
or can be found in the turn_secret table in the database.
This option can be used with long\-term credentials mechanisms only \-
it does not make much sense with the short\-term mechanism.
.TP
.B
\fB\-\-dh566\fP
Use 566 bits predefined DH TLS key. Default size of the key is 1066.
.TP
.B
\fB\-\-dh2066\fP
Use 2066 bits predefined DH TLS key. Default size of the key is 1066.
.TP
.B
\fB\-\-no\-sslv3\fP
Do not allow SSLv3 protocol.
.TP
.B
\fB\-\-no\-tlsv1\fP
Do not allow TLSv1 protocol.
.TP
.B
\fB\-\-no\-tlsv1_1\fP
Do not allow TLSv1.1 protocol.
.TP
.B
\fB\-\-no\-tlsv1_2\fP
Do not allow TLSv1.2 protocol.
.TP
.B
\fB\-\-no\-udp\fP
Do not start UDP client listeners.
.TP
.B
\fB\-\-no\-tcp\fP
Do not start TCP client listeners.
.TP
.B
\fB\-\-no\-tls\fP
Do not start TLS client listeners.
.TP
.B
\fB\-\-no\-dtls\fP
Do not start DTLS client listeners.
.TP
.B
\fB\-\-no\-udp\-relay\fP
Do not allow UDP relay endpoints defined in RFC 5766, 
use only TCP relay endpoints as defined in RFC 6062.
.TP
.B
\fB\-\-no\-tcp\-relay\fP
Do not allow TCP relay endpoints defined in RFC 6062, 
use only UDP relay endpoints as defined in RFC 5766. 
.TP
.B
\fB\-\-stale\-nonce\fP
Use extra security with nonce value having limited lifetime (600 secs). 
.TP
.B
\fB\-\-no\-stdout\-log\fP
Flag to prevent stdout log messages.
By default, all log messages are going to both stdout and to
the configured log file. With this option everything will be going to 
the log file only (unless the log file itself is stdout).
.TP
.B
\fB\-\-syslog\fP
With this flag, all log will be redirected to the system log (syslog).
.TP
.B
\fB\-\-simple\-log\fP
This flag means that no log file rollover will be used, and the log file
name will be constructed as\-is, without PID and date appendage.
This option can be used, for example, together with the logrotate tool.
.TP
.B
\fB\-\-secure\-stun\fP
Require authentication of the STUN Binding request.
By default, the clients are allowed anonymous access to the STUN Binding functionality.
.TP
.B
\fB\-S\fP, \fB\-\-stun\-only\fP
Run as STUN server only, all TURN requests will be ignored. 
Option to suppress TURN functionality, only STUN requests will be processed.
.TP
.B
\fB\-\-no\-stun\fP
Run as TURN server only, all STUN requests will be ignored. 
Option to suppress STUN functionality, only TURN requests will be processed.
.TP
.B
\fB\-\-no\-loopback\-peers\fP
Disallow peers on the loopback addresses (127.x.x.x and ::1).
.TP
.B
\fB\-\-no\-multicast\-peers\fP
Disallow peers on well\-known broadcast addresses 
(224.0.0.0 and above, and FFXX:*).
.TP
.B
\fB\-\-sha256\fP
Require SHA256 digest function to be used for the message integrity.
By default, the server uses SHA1 hashes. With this option, the server 
requires the stronger SHA256 hashes. The client application must support
SHA256 hash function if this option is used. If the server obtains a message 
from the client with a weaker (SHA1) hash function then the server returns 
error code 426.
.TP
.B
\fB\-\-mobility\fP
Mobility with ICE (MICE) specs support.
.TP
.B
\fB\-\-no\-cli\fP
Turn OFF the CLI support. By default it is always ON.
See also \fIoptions\fP \fB\-\-cli\-ip\fP and \fB\-\-cli\-port\fP.
.TP
.B
\fB\-\-server\-relay\fP
Server relay. NON\-STANDARD AND DANGEROUS OPTION. 
Only for those applications when we want to run 
server applications on the relay endpoints.
This option eliminates the IP permissions check 
on the packets incoming to the relay endpoints.
See http://tools.ietf.org/search/rfc5766#section\-17.2.3 .
.TP
.B
\fB\-\-udp\-self\-balance\fP
(recommended for older Linuxes only)
Automatically balance UDP traffic over auxiliary servers
(if configured). The load balancing is using the 
ALTERNATE\-SERVER mechanism. The TURN client must support 
300 ALTERNATE\-SERVER response for this functionality.
.TP
.B
\fB\-h\fP
Help.
.TP
.B
Options with required values:
.TP
.B
\fB\-d\fP, \fB\-\-listening\-device\fP
Listener interface device.
(NOT RECOMMENDED. Optional functionality, Linux only). 
The \fIturnserver\fP process must have root privileges to bind the 
listening endpoint to a device. If \fIturnserver\fP must run as a 
process without root privileges, then just do not use this setting.
.TP
.B
\fB\-L\fP, \fB\-\-listening\-ip\fP
Listener IP address of relay server. 
Multiple listeners can be specified, for example:
\fB\-L\fP ip1 \fB\-L\fP ip2 \fB\-L\fP ip3
If no \fBIP\fP(s) specified, then all IPv4 and 
IPv6 system IPs will be used for listening.
The same \fBip\fP(s) can be used as both listening and relay \fBip\fP(s).
.TP
.B
\fB\-p\fP, \fB\-\-listening\-port\fP
TURN listener port for UDP and TCP listeners (Default: 3478).
Note: actually, TLS & DTLS sessions can connect to the "plain" TCP & UDP
\fBport\fP(s), too \- if allowed by configuration.
.TP
.B
\fB\-\-tls\-listening\-port\fP
TURN listener port for TLS and DTLS listeners (Default: 5349).
Note: actually, "plain" TCP & UDP sessions can connect to the TLS & DTLS
\fBport\fP(s), too \- if allowed by configuration. The TURN server 
"automatically" recognizes the type of traffic. Actually, two listening
endpoints (the "plain" one and the "tls" one) are equivalent in terms of
functionality; but we keep both endpoints to satisfy the RFC 5766 specs.
For secure TCP connections, we currently support SSL version 3 and 
TLS versions 1.0, 1.1, 1.2.
For secure UDP connections, we support DTLS version 1.
.TP
.B
\fB\-\-alt\-listening\-port\fP
Alternative listening port for UDP and TCP listeners;
default (or zero) value means "listening port plus one".
This is needed for STUN CHANGE_REQUEST \- in RFC 5780 sense
or in old RFC 3489 sense \- for NAT behavior discovery). The \fBTURN Server\fP
supports CHANGE_REQUEST only if it is started with more than one
listening IP address of the same family (IPv4 or IPv6). The CHANGE_REQUEST
is only supported by UDP protocol, other protocols are listening
on that endpoint only for "symmetry".
.TP
.B
\fB\-\-alt\-tls\-listening\-port\fP
Alternative listening port for TLS and DTLS protocols.
Default (or zero) value means "TLS listening port plus one".
.TP
.B
\fB\-\-aux\-server\fP
Auxiliary STUN/TURN server listening endpoint.
Aux servers have almost full TURN and STUN functionality.
The (minor) limitations are:
.RS
.IP 1) 4
Auxiliary servers do not have alternative ports and
they do not support STUN RFC 5780 functionality (CHANGE REQUEST).
.IP 2) 4
Auxiliary servers also are never returning ALTERNATIVE\-SERVER reply.
.RE
.PP
Valid formats are 1.2.3.4:5555 for IPv4 and [1:2::3:4]:5555 for IPv6.
There may be multiple aux\-server \fIoptions\fP, each will be used for listening
to client requests.
.TP
.B
\fB\-i\fP, \fB\-\-relay\-device\fP
Relay interface device for relay sockets 
(NOT RECOMMENDED. Optional, Linux only).
.TP
.B
\fB\-E\fP, \fB\-\-relay\-ip\fP
Relay address (the local IP address that 
will be used to relay the packets to the 
peer). Multiple relay addresses may be used:
\fB\-E\fP ip1 \fB\-E\fP ip2 \fB\-E\fP ip3
The same \fBIP\fP(s) can be used as both listening \fBIP\fP(s) and relay \fBIP\fP(s).
If no relay \fBIP\fP(s) specified, then the \fIturnserver\fP will apply the 
default policy: it will decide itself which relay addresses to be 
used, and it will always be using the client socket IP address as 
the relay IP address of the TURN session (if the requested relay 
address family is the same as the family of the client socket).
.TP
.B
\fB\-X\fP, \fB\-\-external\-ip\fP
\fBTURN Server\fP public/private address mapping, if the server is behind NAT.
In that situation, if a \fB\-X\fP is used in form "\fB\-X\fP <ip>" then that ip will be reported
as relay IP address of all allocations. This scenario works only in a simple case
when one single relay address is be used, and no CHANGE_REQUEST functionality is 
required. That single relay address must be mapped by NAT to the 'external' IP.
The "external\-ip" value, if not empty, is returned in XOR\-RELAYED\-ADDRESS field.
For that 'external' IP, NAT must forward ports directly (relayed port 12345
must be always mapped to the same 'external' port 12345).
In more complex case when more than one IP address is involved,
that option must be used several times, each entry must
have form "\fB\-X\fP <public\-ip/private\-ip>", to map all involved addresses.
CHANGE_REQUEST (RFC5780 or RFC3489) NAT discovery STUN functionality will work 
correctly, if the addresses are mapped properly, even when the TURN server itself 
is behind A NAT.
By default, this value is empty, and no address mapping is used.
.TP
.B
\fB\-m\fP, \fB\-\-relay\-threads\fP
Number of relay threads to handle the established connections
(in addition to authentication thread and the listener thread).
If explicitly set to 0 then application runs relay process in a single thread,
in the same thread with the listener process (the authentication thread will 
still be a separate thread). If not set then the default threading policy
will be employed. In older systems (before Linux kernel 3.9),
the number of UDP threads is always one threads per network listening endpoint \-
unless "\fB\-m\fP 0" or "\fB\-m\fP 1" is set.
.TP
.B
\fB\-\-min\-port\fP
Lower bound of the UDP port range for relay 
endpoints allocation.
Default value is 49152, according to RFC 5766.
.TP
.B
\fB\-\-max\-port\fP
Upper bound of the UDP port range for relay 
endpoints allocation.
Default value is 65535, according to RFC 5766.
.TP
.B
\fB\-u\fP, \fB\-\-user\fP
Long\-term security mechanism credentials user account, 
in the column\-separated form username:key. 
Multiple user accounts may used in the command line.
The key is either the user password, or
the key is generated
by \fIturnadmin\fP command. In the second case,
the key must be prepended with 0x symbols.
The key is calculated over the user name, 
the realm, and the user password.
This setting may not be used with TURN REST API or
with short\-term credentials mechanism.
.TP
.B
\fB\-r\fP, \fB\-\-realm\fP
Realm to be used for all users. Must be used with long\-term credentials mechanism 
or with TURN REST API.
.TP
.B
\fB\-C\fP, \fB\-\-rest\-api\-separator\fP
This is the timestamp/username separator symbol (character) in TURN REST API.
The default value is :.
.TP
.B
\fB\-q\fP, \fB\-\-user\-quota\fP
Per\-user allocations quota: how many concurrent 
allocations a user can create.
.TP
.B
\fB\-Q\fP, \fB\-\-total\-quota\fP
Total allocations quota: global limit on concurrent allocations.
.TP
.B
\fB\-\-static\-auth\-secret\fP
Static authentication secret value (a string) for TURN REST API only.
If not set, then the turn server will try to use the dynamic value 
in turn_secret table in user database (if present). The database\-stored
value can be changed on\-the\-fly by a separate program, so this is why
that other mode is dynamic. Multiple shared secrets can be used
(both in the database and in the "static" fashion).
.TP
.B
\fB\-s\fP, \fB\-\-max\-bps\fP
Max bytes\-per\-second bandwidth a TURN session is allowed to handle
(input and output network streams are treated separately). Anything above 
that limit will be dropped or temporary suppressed (within the
available buffer limits).
.TP
.B
\fB\-\-cert\fP
Certificate file, PEM format. Same file 
search rules applied as for the configuration 
file. If both \fB\-\-no\-tls\fP and \fB\-\-no\-dtls\fP \fIoptions\fP 
are specified, then this parameter is not needed.
Default value is turn_server_cert.pem.
.TP
.B
\fB\-\-pkey\fP
Private key file, PEM format. Same file 
search rules applied as for the configuration 
file. If both \fB\-\-no\-tls\fP and \fB\-\-no\-dtls\fP \fIoptions\fP 
are specified, then this parameter is not needed.
Default value is turn_server_pkey.pem.
.TP
.B
\fB\-\-pkey\-pwd\fP
If the private key file is encrypted, then this password to be used.
.TP
.B
\fB\-\-cipher\-list\fP
Allowed OpenSSL cipher list for TLS/DTLS connections.
Default value is "DEFAULT".
.TP
.B
\fB\-\-CA\-file\fP
CA file in OpenSSL format. 
Forces TURN server to verify the client SSL certificates.
By default, no CA is set and no client certificate check is performed.
.TP
.B
\fB\-\-ec\-curve\-name\fP
Curve name for EC ciphers, if supported by OpenSSL library (TLS and DTLS).
The default value is prime256v1.
.TP
.B
\fB\-\-dh\-file\fP
Use custom DH TLS key, stored in PEM format in the file.
Flags \fB\-\-dh566\fP and \fB\-\-dh2066\fP are ignored when the DH key is taken from a file.
.TP
.B
\fB\-l\fP, \fB\-\-log\-file\fP
Option to set the full path name of the log file.
By default, the \fIturnserver\fP tries to open a log file in 
/var/log/\fIturnserver\fP, /var/log, /var/tmp, /tmp and . (current) 
directories (which file open operation succeeds 
first that file will be used). With this option you can set the 
definite log file name.
The special names are "stdout" and "\-" \- they will force everything 
to the stdout. Also, "syslog" name will redirect everything into
the system log (syslog), as if the option "\fB\-\-syslog\fP" was set. 
In the runtime, the logfile can be reset with the SIGHUP signal 
to the \fIturnserver\fP process.
.TP
.B
\fB\-\-alternate\-server\fP
Option to set the "redirection" mode. The value of this option
will be the address of the alternate server for UDP & TCP service in form of 
<ip>[:<port>]. The server will send this value in the attribute
ALTERNATE\-SERVER, with error 300, on ALLOCATE request, to the client.
Client will receive only values with the same address family
as the client network endpoint address family. 
See RFC 5389 and RFC 5766 for ALTERNATE\-SERVER functionality description. 
The client must use the obtained value for subsequent TURN communications.
If more than one \fB\-\-alternate\-server\fP \fIoptions\fP are provided, then the functionality
can be more accurately described as "load\-balancing" than a mere "redirection". 
If the port number is omitted, then the default port 
number 3478 for the UDP/TCP protocols will be used.
Colon (:) characters in IPv6 addresses may conflict with the syntax of 
the option. To alleviate this conflict, literal IPv6 addresses are enclosed 
in square brackets in such resource identifiers, for example: 
[2001:db8:85a3:8d3:1319:8a2e:370:7348]:3478 . 
Multiple alternate servers can be set. They will be used in the
round\-robin manner. All servers in the pool are considered of equal weight and 
the load will be distributed equally. For example, if we have 4 alternate servers, 
then each server will receive 25% of ALLOCATE requests. An alternate TURN server 
address can be used more than one time with the alternate\-server option, so this 
can emulate "weighting" of the servers. 
.TP
.B
\fB\-\-tls\-alternate\-server\fP
Option to set alternative server for TLS & DTLS services in form of 
<ip>:<port>. If the port number is omitted, then the default port 
number 5349 for the TLS/DTLS protocols will be used. See the previous option for the 
functionality description.
.TP
.B
\fB\-O\fP, \fB\-\-redis\-statsdb\fP
Redis status and statistics database connection string, if used (default \- empty, 
no Redis stats DB used). This database keeps allocations status information, and it can 
be also used for publishing and delivering traffic and allocation event notifications.
This database option can be used independently of \fB\-\-redis\-userdb\fP option,
and actually Redis can be used for status/statistics and MySQL or PostgreSQL can
be used for the user database.
The connection string has the same parameters as redis\-userdb connection string.
.TP
.B
\fB\-\-max\-allocate\-timeout\fP
Max time, in seconds, allowed for full allocation establishment. 
Default is 60 seconds.
.PP
\fB\-\-denied\-peer\-ip\fP=<IPaddr[\fB\-IPaddr\fP]>
.PP
\fB\-\-allowed\-peer\-ip\fP=<IPaddr[\fB\-IPaddr\fP]> Options to ban or allow specific ip addresses or ranges 
of ip addresses. If an ip address is specified as both allowed and denied, then 
the ip address is considered to be allowed. This is useful when you wish to ban
a range of ip addresses, except for a few specific ips within that range.
This can be used when you do not want users of the turn server to be able to access
machines reachable by the turn server, but would otherwise be unreachable from the 
internet (e.g. when the turn server is sitting behind a NAT). The 'white" and "black" peer 
IP ranges can also be dynamically changed in the database. 
The allowed/denied addresses (white/black lists) rules are very simple:
.RS
.IP 1) 4
If there is no rule for an address, then it is allowed;
.IP 2) 4
If there is an "allowed" rule that fits the address then it is allowed \- no matter what;
.IP 3) 4
If there is no "allowed" rule that fits the address, and if there is a "denied" rule that
fits the address, then it is denied.
.RE
.TP
.B
\fB\-\-pidfile\fP
File name to store the pid of the process.
Default is /var/run/turnserver.pid (if superuser account is used) or
/var/tmp/turnserver.pid .
.TP
.B
\fB\-\-proc\-user\fP
User name to run the process. After the initialization, the \fIturnserver\fP process
will make an attempt to change the current user ID to that user.
.TP
.B
\fB\-\-proc\-group\fP
Group name to run the process. After the initialization, the \fIturnserver\fP process
will make an attempt to change the current group ID to that group.
.TP
.B
\fB\-\-cli\-ip\fP
Local system IP address to be used for CLI management interface.
The \fIturnserver\fP process can be accessed for management with telnet,
at this IP address and on the CLI port (see the next parameter). 
Default value is 127.0.0.1. You can use telnet or putty (in telnet mode)
to access the CLI management interface. 
.TP
.B
\fB\-\-cli\-port\fP
CLI management interface listening port. Default is 5766.
.TP
.B
\fB\-\-cli\-password\fP
CLI access password. Default is empty (no password).
.TP
.B
\fB\-\-cli\-max\-output\-sessions\fP
Maximum number of output sessions in ps CLI command.
This value can be changed on\-the\-fly in CLI. The default value is 256.
.TP
.B
\fB\-\-ne\fP=[1|2|3]

ChangeLog
5/29/2015 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.5.8 'Marshal West':
	- The fix for Issue 11 backported from coturn;
	
3/18/2015 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.5.7 'Marshal West':
	- plain passwords in Redis database fixed;

2/28/2015 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.5.6 'Marshal West':
	- bandwidth control fixed;
	- STUN/TURN control traffic counted separately from data traffic,
	for the sake of the bandwidth control;
	- REST API docs fixed;

2/3/2015 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.5.5 'Marshal West':
	- hostname-to-ip-address resolution fixed;
	- TRANSPORT attribute handling fixed;
	- TOS (DiffServer) and TTL IP header field handling fixed;
	- peer logging added;

1/26/2015 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.5.4 'Marshal West':
	- SSLv2 support removed;
	- OpenSSL 1.1.0 support added.
	
12/14/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.5.3 'Marshal West':
	- Redis read message queue backlog fixed.
	- Typos fixed in the docs.
	
11/10/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.5.2 'Marshal West':
	- TLS connection procedure improved in uclient test program.
	- Reliability improvements (Issue 141).
	- external address mapping fixes for Amazon EC2.
	- docs fixes.

11/07/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.5.1 'Marshal West':
	- Request re-transmission implemented in uclient test program.
	- TLS connection procedure improved in uclient test program.

10/31/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.4.6 'Marshal West':
	- Relay RTCP sockets ports allocation fixed.
	- SSL renegotiation callback fixed.
	- List of libraries cleaned.
	- Debian UFW file added to the Debian package.

10/16/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.4.5 'Marshal West':
	- Mobile sessions authentication improved.
	- Authentication code cleaning.
	- Quota allocation fixed.
	- Core dump fixed.
	- Memory code cleaning.
	- Logging code cleaning.
	- TCP/TLS tests improved.
	- relay RTCP sockets ports allocation fixed.
	
09/05/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.4.4 'Marshal West':
	- TLS renegotiation DoS attack prevention implemented (Issue 136);
	- FQDN is allowed as a relay-ip or a listener-ip parameter
	(patch provided by Iñaki Baz Castillo).
	
09/01/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.4.3 'Marshal West':
	- Redis DB user key database operation fixed.
	- SHA-256 memory leak fixed.
	- Mobility ticket retransmission fixed.
	
08/14/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.4.2 'Marshal West':
	- MySQL SSL connection support.
	- Crash fixed when incorrect DB connection string.
	- Minor docs fixes.
	
07/29/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.4.1 'Marshal West':
	- Server-side forceful session cancellation implemented
	(in telnet console).
	
07/18/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.96 'Marshal West':
	- TCP 6062 relay connection fix.
	- "pu" CLI command fixed.
	- Some crash conditions fixed.
	
07/11/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.95 'Marshal West':
	- working on compilation warnings.
	- Allocation mismatch condition fixed.
	- STUN BINDING response fixed in the case of -X (external address) option.
	
06/13/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.94 'Marshal West':
	- Redis DB connection status fixed (Issue 129).
	- SIGHUP for logfile reset implemented
	(Gustavo Garcia suggestion).
	- log reset command in CLI added.
	- Some error code corrections:
	    * "Mobility forbidden" error changed, to value 405.
	    * "Wrong credentials" situation is now treated as error 441.

06/03/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.92 'Marshal West':
	-Issues 126, 127 and 128 fixed.

06/02/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.91 'Marshal West':
	-Issue 128 (sigfault in RFC 6062 TCP operations) fixed.

05/30/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.9 'Marshal West':
	- ephemeral REST API credentials fixed in RFC 6062 TCP communications
	(Issue 126 and Issue 127).
	- Auth code cleaning.
	- Password change in the database of an active session is supported.
	
05/07/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.8 'Marshal West':
	- Code cleaning (compilation warnings, etc).
	- TLS/DTLS log messages fixed.
	- uclient performance improvements.
	- channel function place holders  defined for kernel-level calls.

04/23/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.7 'Marshal West':
	- Performance and socket logging fix (improved) (Issue 123).
	- Code cleaning.
	
04/13/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.6 'Marshal West':
	- Addresses logging fixed.
	- Redis admin options fixed.
	- Redis compilation cleaned.
	- Performance and logging socket problem fixed (Issue 123).

04/07/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.5 'Marshal West':
	- Mobile allocation quota fixed (issue 121);
	- --simple-log option added (Issue 122);
	- documentation fixes (REST API, Redis).

04/06/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.4 'Marshal West':
	- Mobile TCP sessions fixed (issue 120);
	- log information improvements.

04/04/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.3 'Marshal West':
	- Pkey and cert file descriptors to be closed 
	on initialization (issue 118);
	- Address bind indefinite cycle on start-up fixed
	(Issue 119);
	- Allocation counters time lag improved.
	
03/30/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.2 'Marshal West':
	- Allocation counters fixed (issue 117);
	- a possible core dump in the server code fixed;
	- a possible memory leak in server fixed.
	
03/29/2014 Oleg Moskalenko <mom040267@gmail.com>
Version 3.2.3.1 'Marshal West':
	- TCP congestion avoidance completed.
	- Read and write streams are treated separately in
	bandwidth control.
	- Test client fixed.
	- Experimental SHA256 key storage supported.


rpm/turnserver.init.el
#!/bin/bash
#
# Startup script for TURN Server
#
# chkconfig: 345 85 15
# description: RFC 5766 TURN Server
#
# processname: turnserver
# pidfile: /var/run/turnserver.pid
# config: /etc/turnserver/turnserver.conf
#
### BEGIN INIT INFO
# Provides: turnserver
# Required-Start: $local_fs $network
# Short-Description: RFC 5766 TURN Server
# Description: RFC 5766 TURN Server
### END INIT INFO

# Source function library.
. /etc/rc.d/init.d/functions

TURN=/usr/bin/turnserver
PROG=turnserver
TURNCFG=/etc/turnserver/$PROG.conf
PID_FILE=/var/run/$PROG.pid
LOCK_FILE=/var/lock/subsys/$PROG
DEFAULTS=/etc/sysconfig/$PROG
RETVAL=0
USER=turnserver

start() {
	echo -n $"Starting $PROG: "
	daemon --user=$USER $TURN $OPTIONS
	RETVAL=$?
	if [ $RETVAL = 0 ]; then
		pidofproc $TURN > $PID_FILE
		RETVAL=$?
		[ $RETVAL = 0 ] && touch $LOCK_FILE && success
	fi
	echo
	return $RETVAL
}

stop() {
	echo -n $"Stopping $PROG: "
	killproc $TURN
	RETVAL=$?
	echo
	[ $RETVAL = 0 ] && rm -f $LOCK_FILE $PID_FILE
}

[ -f $DEFAULTS ] && . $DEFAULTS
OPTIONS="-c $TURNCFG -o --no-stdout-log --syslog $EXTRA_OPTIONS"

# See how we were called.
case "$1" in
	start)
		start
		;;
	stop)
		stop
		;;
	status)
		status $TURN
		RETVAL=$?
		;;
	restart)
		stop
		start
		;;
	condrestart)
		if [ -f $PID_FILE ] ; then
			stop
			start
		fi
		;;
	*)
		echo $"Usage: $PROG {start|stop|restart|condrestart|status|help}"
		exit 1
esac

exit $RETVAL
